File: Algorithm/Python/649. Dota2 Senate.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2017-08-08 19:48:52
# @Last Modified by:   WuLC
# @Last Modified time: 2017-08-08 19:53:44


# greedy, each senator ban its next opponent senator
from collections import deque
class Solution(object):
    def predictPartyVictory(self, senate):
        """
        :type senate: str
        :rtype: str
        """
        r, d = deque(), deque()
        n = len(senate)
        for i in xrange(len(senate)):
            if senate[i] ==  'R':
                r.append(i)
            else:
                d.append(i)
            
        while len(r) > 0 and len(d) > 0:
            if r[0] > d[0]:
                d.append(d.popleft() + n)
                r.popleft()
            else:
                r.append(r.popleft() + n)
                d.popleft()
        return 'Radiant' if len(r) > 0 else 'Dire'
                    

File: Algorithm/Java/649. Dota2 Senate.java
/*
 * Created on Sun Apr 08 2018 18:28:56
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */


// use two queues, greedy
// pay attention that Queue is a interface in Java, so initialize it with LinkedList
class Solution 
{
    public String predictPartyVictory(String senate) 
    {
        Queue<Integer> rq = new LinkedList<Integer>();
        Queue<Integer> dq = new LinkedList<Integer>();
        char[] cs = senate.toCharArray();
        for(int i=0; i<cs.length; i++)
        {
            if (cs[i] == 'R') rq.offer(i);
            else dq.offer(i);
        }
        int n = cs.length;
        while(dq.size() > 0 && rq.size() > 0)
        {
            if (dq.peek() < rq.peek())
            {
                rq.poll();
                dq.offer(dq.poll() + n);
            }
            else
            {
                dq.poll();
                rq.offer(rq.poll() + n);
            }
        }
        if (dq.size() > 0) return "Dire";
        else return "Radiant";
    }
}